pkg/ext-proc/main.go

```diff
@@ -144,53 +157,103 @@ func main() {
 		Zone:           *zone,
 		ServerPoolName: *serverPoolName,
 	}).SetupWithManager(mgr); err != nil {
-		klog.Error(err, "Error setting up EndpointSliceReconciler")
+		klog.Fatalf("Error setting up EndpointSliceReconciler: %v", err)
 	}
 
-	errChan := make(chan error)
+	// Start the controller manager
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
+
 		if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-			klog.Error(err, "Error running manager")
-			errChan <- err
+			errChan <- fmt.Errorf("controller manager failed to start: %w", err)
 		}
+		klog.Info("controller manager started")
 	}()
 
-	s := grpc.NewServer()
+	// Start the health server
+	var healthSvr *grpc.Server
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
 
-	pp := backend.NewProvider(&vllm.PodMetricsClientImpl{}, datastore)
-	if err := pp.Init(*refreshPodsInterval, *refreshMetricsInterval); err != nil {
-		klog.Fatalf("failed to initialize: %v", err)
-	}
-	extProcPb.RegisterExternalProcessorServer(
-		s,
-		handlers.NewServer(
-			pp,
-			scheduling.NewScheduler(pp),
-			*targetPodHeader,
-			datastore))
-	healthPb.RegisterHealthServer(s, &healthServer{})
-
-	klog.Infof("Starting gRPC server on port :%v", *port)
-
-	// shutdown
-	var gracefulStop = make(chan os.Signal, 1)
-	signal.Notify(gracefulStop, syscall.SIGTERM)
-	signal.Notify(gracefulStop, syscall.SIGINT)
+		healthSvr = grpc.NewServer()
+		healthPb.RegisterHealthServer(healthSvr, &healthServer{datastore: datastore})
+		healthLis, err := net.Listen("tcp", fmt.Sprintf(":%d", *grpcHealthPort))
+		if err != nil {
+			errChan <- fmt.Errorf("health server failed to listen: %w", err)
+			return
+		}
+
+		if err := healthSvr.Serve(healthLis); err != nil && err != grpc.ErrServerStopped {
+			errChan <- fmt.Errorf("health server failed: %w", err)
+		}
+		klog.Infof("Health server serving on port: %d", *grpcHealthPort)
+	}()
+
+	// Start the ext-proc server
+	var extSvr *grpc.Server
+	wg.Add(1)
 	go func() {
-		select {
-		case sig := <-gracefulStop:
-			klog.Infof("caught sig: %+v", sig)
-			os.Exit(0)
-		case err := <-errChan:
-			klog.Infof("caught error in controller: %+v", err)
-			os.Exit(0)
+		defer wg.Done()
+
+		extSvr = grpc.NewServer()
+		lis, err := net.Listen("tcp", fmt.Sprintf(":%d", *grpcPort))
+		if err != nil {
+			errChan <- fmt.Errorf("ext-proc server failed to listen: %w", err)
+			return
+		}
+
+		pp := backend.NewProvider(&vllm.PodMetricsClientImpl{}, datastore)
+		if err := pp.Init(*refreshPodsInterval, *refreshMetricsInterval); err != nil {
+			errChan <- fmt.Errorf("failed to initialize backend provider: %w", err)
+			return
 		}
+		extProcPb.RegisterExternalProcessorServer(
+			extSvr,
+			handlers.NewServer(
+				pp,
+				scheduling.NewScheduler(pp),
+				*targetPodHeader,
+				datastore,
+			),
+		)
 
+		if err := extSvr.Serve(lis); err != nil && err != grpc.ErrServerStopped {
```

Comment:
Please add a comment clarifying that this is blocking and will return when shutting down.